### PR TITLE
Add confluent schema registry message decoder

### PR DESCRIPTION
**Pull request by elonazoulay**
Add a message decoder that can read kafka messages produced using confluent schema registry. The schema id is stored in the message instead of the entire schema and can be retrieved from the schema registry.

### CODE REVIEW

File: pinot-plugins/pinot-input-format/pinot-avro/pom.xml [hunk 1]
```diff
@@ -33,6 +33,7 @@
   <url>https://pinot.apache.org/</url>
   <properties>
     <pinot.root>${basedir}/../../..</pinot.root>
+    <kafka.lib.version>2.0.0</kafka.lib.version>
```
Comment (mcvsubbu) [inline pinot-plugins/pinot-input-format/pinot-avro/pom.xml]: we should not have pinot-avro pull in kafka

File: pinot-plugins/pinot-input-format/pinot-avro/src/main/java/org/apache/pinot/plugin/inputformat/avro/KafkaConfluentSchemaRegistryAvroMessageDecoder.java [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+import org.apache.pinot.spi.plugin.PluginManager;
+import org.apache.pinot.spi.stream.StreamMessageDecoder;
+
+import java.util.Arrays;
+import java.util.Map;
+
+import static com.google.common.base.Preconditions.checkState;
+import static java.util.Objects.requireNonNull;
+
+public class KafkaConfluentSchemaRegistryAvroMessageDecoder implements StreamMessageDecoder<byte[]> {
```
Comment (mcvsubbu) [inline pinot-plugins/pinot-input-format/pinot-avro/src/main/java/org/apache/pinot/plugin/inputformat/avro/KafkaConfluentSchemaRegistryAvroMessageDecoder.java]: Please move this to kafka plugins, see `KafkaAvroMessageDecoder`

**Comment by kishoreg** [comment]
This should be a separate plugin - pinot-confluent-avro. I am guessing pinot-kafka module should work with confluent kafka as well.

File: pinot-plugins/pinot-input-format/pinot-avro/src/main/java/org/apache/pinot/plugin/inputformat/avro/KafkaConfluentSchemaRegistryAvroMessageDecoder.java (hunk 2 above)
Comment (elonazoulay) [reply pinot-plugins/pinot-input-format/pinot-avro/src/main/java/org/apache/pinot/plugin/inputformat/avro/KafkaConfluentSchemaRegistryAvroMessageDecoder.java]: This is in the same package as KafkaAvroMessageDecoder - do you mean have it in a separate module under pinot-input-format?

**Comment by elonazoulay** [comment]
Hi @mcvsubbu, @kishoreg  - can you take another pass when you have a chance?

File: pinot-plugins/pinot-input-format/pinot-confluent-avro/src/main/java/org/apache/pinot/plugin/inputformat/avro/confluent/KafkaConfluentSchemaRegistryAvroMessageDecoder.java [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+import org.apache.pinot.spi.plugin.PluginManager;
+import org.apache.pinot.spi.stream.StreamMessageDecoder;
+
+import java.util.Arrays;
+import java.util.Map;
+
+import static com.google.common.base.Preconditions.checkState;
+import static java.util.Objects.requireNonNull;
+
+public class KafkaConfluentSchemaRegistryAvroMessageDecoder implements StreamMessageDecoder<byte[]> {
```
Comment (kishoreg) [inline pinot-plugins/pinot-input-format/pinot-confluent-avro/src/main/java/org/apache/pinot/plugin/inputformat/avro/confluent/KafkaConfluentSchemaRegistryAvroMessageDecoder.java]: can you add some comments on how this is different from AvroDecoder we have in pinot-avro or point to some external link

File: pinot-plugins/pinot-input-format/pinot-confluent-avro/src/main/java/org/apache/pinot/plugin/inputformat/avro/confluent/KafkaConfluentSchemaRegistryAvroMessageDecoder.java [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+
+    @Override
+    public void init(Map<String, String> props, Schema indexingSchema, String topicName) throws Exception {
+        checkState(props.containsKey(SCHEMA_REGISTRY_REST_URL), "Missing required property '%s'", SCHEMA_REGISTRY_REST_URL);
+        String schemaRegistryUrl = props.get(SCHEMA_REGISTRY_REST_URL);
+        pinotSchema = requireNonNull(indexingSchema, "indexingSchema is null");
+        SchemaRegistryClient schemaRegistryClient = new CachedSchemaRegistryClient(schemaRegistryUrl, 1000);
+        deserializer = new KafkaAvroDeserializer(schemaRegistryClient);
+        this.topicName = requireNonNull(topicName, "topicName is null");
+        this.avroRecordConvetrer = PluginManager.get().createInstance(AvroRecordExtractor.class.getName());
```
Comment (kishoreg) [inline pinot-plugins/pinot-input-format/pinot-confluent-avro/src/main/java/org/apache/pinot/plugin/inputformat/avro/confluent/KafkaConfluentSchemaRegistryAvroMessageDecoder.java]: Its probably ok to duplicate the AvroRecordExtractor. Plugin depending on another plugin will probably make things harder. For e.g. Avro versions might be different between pinot-avro and pinot-confluent-avro

File: pinot-plugins/pinot-input-format/pinot-confluent-avro/src/main/java/org/apache/pinot/plugin/inputformat/avro/confluent/KafkaConfluentSchemaRegistryAvroMessageDecoder.java [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+    }
+
+    @Override
+    public GenericRow decode(byte[] payload, GenericRow destination) {
+        Record avroRecord = (Record) deserializer.deserialize(topicName, payload);
+        return avroRecordConvetrer.extract(pinotSchema, avroRecord, destination);
```
Comment (kishoreg) [inline pinot-plugins/pinot-input-format/pinot-confluent-avro/src/main/java/org/apache/pinot/plugin/inputformat/avro/confluent/KafkaConfluentSchemaRegistryAvroMessageDecoder.java]: typo avroRecordConvetrer

File: pinot-plugins/pinot-input-format/pinot-confluent-avro/src/main/java/org/apache/pinot/plugin/inputformat/avro/confluent/KafkaConfluentSchemaRegistryAvroMessageDecoder.java (hunk 4 above)
Comment (kishoreg) [reply pinot-plugins/pinot-input-format/pinot-confluent-avro/src/main/java/org/apache/pinot/plugin/inputformat/avro/confluent/KafkaConfluentSchemaRegistryAvroMessageDecoder.java]: Another option is to create a separate module(not a plugin) called pinot-avro-utils let both pinot-avro and pinot-confluent-avro share that.

File: pom.xml [hunk 6]
```diff
@@ -138,6 +138,7 @@
     To change kafka connector dependency, we only need to update this version number config.
     TODO: figure out a way to inject kafka dependency instead of explicitly setting the kafka module dependency -->
     <kafka.version>2.0</kafka.version>
+    <confluent.version>5.3.1</confluent.version>
```
Comment (kishoreg) [inline pom.xml]: should we move this to the pom specific to this plugin.

File: pom.xml [hunk 7]
```diff
@@ -212,6 +213,14 @@
       </build>
     </profile>
   </profiles>
+
```
Comment (kishoreg) [inline pom.xml]: move to pom specific to this plugin?

File: pinot-plugins/pinot-input-format/pinot-confluent-avro/src/main/java/org/apache/pinot/plugin/inputformat/avro/confluent/KafkaConfluentSchemaRegistryAvroMessageDecoder.java (hunk 4 above)
Comment (elonazoulay) [reply pinot-plugins/pinot-input-format/pinot-confluent-avro/src/main/java/org/apache/pinot/plugin/inputformat/avro/confluent/KafkaConfluentSchemaRegistryAvroMessageDecoder.java]: Created pinot-avro-base that both pinot-avro and pinot-confluent-avro plugins use.

File: pinot-plugins/pinot-input-format/pinot-csv/pom.xml [hunk 8]
```diff
@@ -43,5 +43,11 @@
       <groupId>commons-lang</groupId>
       <artifactId>commons-lang</artifactId>
     </dependency>
+    <dependency>
```
Comment (fx19880617) [inline pinot-plugins/pinot-input-format/pinot-csv/pom.xml]: is this needed?

File: pinot-plugins/pinot-input-format/pinot-json/pom.xml [hunk 9]
```diff
@@ -39,5 +39,11 @@
       <groupId>commons-lang</groupId>
       <artifactId>commons-lang</artifactId>
     </dependency>
+    <dependency>
+      <groupId>org.apache.pinot</groupId>
+      <artifactId>pinot-spi</artifactId>
```
Comment (fx19880617) [inline pinot-plugins/pinot-input-format/pinot-json/pom.xml]: is this needed?

File: pinot-plugins/pinot-input-format/pinot-parquet/pom.xml [hunk 10]
```diff
@@ -55,5 +65,11 @@
         </exclusion>
       </exclusions>
     </dependency>
+    <dependency>
+      <groupId>org.apache.pinot</groupId>
+      <artifactId>pinot-spi</artifactId>
```
Comment (fx19880617) [inline pinot-plugins/pinot-input-format/pinot-parquet/pom.xml]: is this needed?

File: pinot-plugins/pinot-input-format/pinot-confluent-avro/pom.xml [hunk 11]
```diff
@@ -0,0 +1,134 @@
...
+        <exclusion>
+          <groupId>org.mortbay.jetty</groupId>
+          <artifactId>jetty</artifactId>
+        </exclusion>
+      </exclusions>
+    </dependency>
+
+    <dependency>
+      <groupId>org.apache.kafka</groupId>
+      <artifactId>kafka-clients</artifactId>
```
Comment (fx19880617) [inline pinot-plugins/pinot-input-format/pinot-confluent-avro/pom.xml]: why we need `kafka-client` dependency?

**Comment by fx19880617** [comment]
overall looks good, some dependencies need to be sorted out.

File: pinot-plugins/pinot-input-format/pinot-avro-base/pom.xml [hunk 12]
```diff
@@ -0,0 +1,59 @@
...
+    <groupId>org.apache.pinot</groupId>
+    <version>0.3.0-SNAPSHOT</version>
+  </parent>
+  <modelVersion>4.0.0</modelVersion>
+
+  <artifactId>pinot-avro-base</artifactId>
+  <name>Pinot Avro Base</name>
+  <url>https://pinot.apache.org/</url>
+  <properties>
+    <pinot.root>${basedir}/../../..</pinot.root>
```
[the remark this answers is not in the document]
Comment (elonazoulay) [reply pinot-plugins/pinot-input-format/pinot-avro-base/pom.xml]: Thanks @fx19880617! Removed. I noticed that this property is set in all the plugins under pinot-input-format. Does it make sense to remove those as well in a separate commit?

File: pinot-plugins/pinot-input-format/pinot-csv/pom.xml (hunk 8 above)
Comment (elonazoulay) [reply pinot-plugins/pinot-input-format/pinot-csv/pom.xml]: The build failed with a dependency error without it.

File: pinot-plugins/pinot-input-format/pinot-json/pom.xml (hunk 9 above)
Comment (elonazoulay) [reply pinot-plugins/pinot-input-format/pinot-json/pom.xml]: The build failed with a dependency error without it.

File: pinot-plugins/pinot-input-format/pinot-parquet/pom.xml (hunk 10 above)
Comment (elonazoulay) [reply pinot-plugins/pinot-input-format/pinot-parquet/pom.xml]: Same as above.

File: pinot-plugins/pinot-input-format/pinot-confluent-avro/pom.xml (hunk 11 above)
Comment (elonazoulay) [reply pinot-plugins/pinot-input-format/pinot-confluent-avro/pom.xml]: The schema registry client has a dependency and a dependency convergence error is thrown without an explicit dependency.

File: pinot-plugins/pinot-input-format/pinot-avro-base/pom.xml (hunk 12 above)
[the remark this answers is not in the document]
Comment (elonazoulay) [reply pinot-plugins/pinot-input-format/pinot-avro-base/pom.xml]: FYI, I had to add it back as the property is used by the build and fails without it (mycila license error).

File: pinot-plugins/pinot-input-format/pinot-confluent-avro/src/main/java/org/apache/pinot/plugin/inputformat/avro/confluent/KafkaConfluentSchemaRegistryAvroMessageDecoder.java (hunk 3 above)
Comment (elonazoulay) [reply pinot-plugins/pinot-input-format/pinot-confluent-avro/src/main/java/org/apache/pinot/plugin/inputformat/avro/confluent/KafkaConfluentSchemaRegistryAvroMessageDecoder.java]: Added in pluggablestreams.rst

File: pinot-plugins/pinot-input-format/pinot-json/pom.xml (hunk 9 above)
Comment (fx19880617) [reply pinot-plugins/pinot-input-format/pinot-json/pom.xml]: I thought `pinot-spi` is in `pinot-plugins` pom (https://github.com/apache/incubator-pinot/blob/master/pinot-plugins/pom.xml#L50) so you don't need to put it here.
Maybe to rebase from current master?

File: pinot-plugins/pinot-input-format/pinot-confluent-avro/pom.xml (hunk 11 above)
Comment (fx19880617) [reply pinot-plugins/pinot-input-format/pinot-confluent-avro/pom.xml]: Is this lib used anywhere in this decoder? shall we exclude this or explicit specify it here?

Comment (elonazoulay) [reply pinot-plugins/pinot-input-format/pinot-confluent-avro/pom.xml]: It's used by the schema registry client, as it stores schemas in a kafka topic, and uses the kafka client to retrieve and store metadata. It's a direct dependency of schema-registry client. If we exclude it then we get a ClassNotFound exception. So I would say leave it explicitly specified. Let me know what you think.

**Comment by elonazoulay** [comment]
Redid the dependencies, let me know when you have a chance to review, thanks!

**Comment by fx19880617** [comment]
> Redid the dependencies, let me know when you have a chance to review, thanks!

Can you update the PR description.

**Comment by kishoreg** [comment]
This looks good. @elonazoulay can add some description

File: docs/pluggable_streams.rst [hunk 13]
```diff
@@ -230,6 +230,42 @@ Please note:
 #. Config ``stream.kafka.consumer.type`` should be specified as ``LowLevel`` to use partition level consumer. (The use of ``simple`` instead of ``LowLevel`` is deprecated)
 #. Configs ``stream.kafka.zk.broker.url`` and ``stream.kafka.broker.list`` are required under ``tableIndexConfig.streamConfigs`` to provide kafka related information.
 
+Here is another example which is used to read avro messages using the schema id from confluent schema registry.
```
[the remark this answers is not in the document]
Comment (elonazoulay) [reply docs/pluggable_streams.rst]: Will do this shortly, thanks!

**Pull request merged**